### PR TITLE
KAFKA-5818: KafkaStreams state transitions not correct

### PR DESCRIPTION
**Pull request by mjsax**
- need to check that state is CRATED at startup
- some minor test cleanup

### CODE REVIEW

**Comment by mjsax** [comment]
Call for review @bbejeck @dguy @guozhangwang

**Comment by mjsax** [comment]
This should be cherry-picked to 0.11.0 for get into 0.11.0.1 bug fix release to avoid a regression. \cc @dguy

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java [hunk 1]
```diff
@@ -673,9 +673,15 @@ private void checkBrokerVersionCompatibility() throws StreamsException {
     public synchronized void start() throws IllegalStateException, StreamsException {
         log.debug("{} Starting Streams client", logPrefix);
 
+        final State oldState = state();
+        final State runningState = State.RUNNING;
+        // this is a mandatory check and not covered by regular state transition checks (cf. KAFKA-5818)
+        if (!(oldState == State.CREATED)) {
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: Hmm.. what if after the check passed, i.e. between line 681 and 682, the state is then transited to running and then to rebalancing?

Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: I think we can introduce a

```
boolean setStateFromCreated()
```

in which we grab the state lock, and then check that state is `CREATED`, and then call `setState(RUNNING)` assuming that `sychronized` is reentrant.

Comment (dguy) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: +1

Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: Ack.
We can't call `setState` though as we would call the stateListener while we hold the lock...

**Comment by mjsax** [comment]
@guozhangwang @dguy Updated this.

**Review by dguy — APPROVED** [review]
Thanks @mjsax, LGTM

**Comment by dguy** [comment]
Merged to trunk. @mjsax could you please raise another PR for 0.11.0? I tried to cherry-pick but there were conflicts (I did try to resolve, but then had test failures). Thanks

**Pull request closed without merging**